[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

cloud456 opened the pull request:
<!--

Thank you for contributing to SeaTunnel! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

## Contribution Checklist
  - Make sure that the pull request corresponds to a [GITHUB issue](https://github.com/apache/seatunnel/issues).
  - Name the pull request in the form "[Feature] [component] Title of the pull request", where *Feature* can be replaced by `Hotfix`, `Bug`, etc.
  - Minor fixes should be named following this pattern: `[hotfix] [docs] Fix typo in README.md doc`.
-->

### Purpose of this pull request

<!-- Describe the purpose of this pull request. For example: This pull request adds checkstyle plugin.-->

At the Re:invent conference in 2024,AWS announced Amazon Aurora DSQL, a new serverless, distributed SQL database. Up to now, there is no tool that can synchronize to DSQL well.
### Does this PR introduce _any_ user-facing change?

<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released SeaTunnel versions or within the unreleased branches such as dev.
If no, write 'No'.
If you are adding/modifying connector documents, please follow our new specifications: https://github.com/apache/seatunnel/issues/4544.
-->


### How was this patch tested?

<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If you are adding E2E test cases, maybe refer to https://github.com/apache/seatunnel/blob/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-cdc-mysql-e2e/src/test/resources/mysqlcdc_to_mysql.conf, here is a good example.
-->


### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  2. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
  3. Add ci label in [label-scope-conf](https://github.com/apache/seatunnel/blob/dev/.github/workflows/labeler/label-scope-conf.yml)
  4. Add e2e testcase in [seatunnel-e2e](https://github.com/apache/seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/)
  5. Update connector [plugin_config](https://github.com/apache/seatunnel/blob/dev/config/plugin_config)

Commit:
Complete the basic development of dsql sink connector

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

Hisoka-X reviewed (commented):
I found dsql support jdbc. Why not add new catalog and dialect in jdbc connector? Please refer https://github.com/apache/seatunnel/tree/dev/seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog and https://github.com/apache/seatunnel/tree/dev/seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect

cloud456 commented:
> I found dsql support jdbc. Why not add new catalog and dialect in jdbc connector? Please refer https://github.com/apache/seatunnel/tree/dev/seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/catalog and https://github.com/apache/seatunnel/tree/dev/seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect

The reason for not using JDBC is that the connection password for DSQL requires generating a temporary token to serve as the password. This token is only valid for 15 minutes by default, therefore requiring token renewal and dataSource reset within that 15-minute timeframe. Implementing this mechanism within JDBC would, in my opinion, be an intrusive behavior and not standard practice. Please refer https://docs.aws.amazon.com/aurora-dsql/latest/userguide/SECTION_authentication-token.html

Hisoka-X commented:
JDBC connector has many features available out of the box and is easier to maintain.

> requires generating a temporary token to serve as the password.

So I think we can add new method named `refreshPassword` in `JdbcConnectionProvider` to adapt this situation.

cloud456 commented:
Glad to hear it.Do you have any schedule for this? Just to mention, it's not only about implementing refreshPassword in JdbcConnectionProvider, but also redesigning JdbcConnectionConfig. DSQL doesn't require a username and password; instead, it creates a password using ak/sk or profile combined with region and cluster endpoint. Therefore, its parameter requirements for JdbcConnectionConfig are different, potentially changing the original parameter requirements, such as making username and password mandatory in the original design

Hisoka-X commented:
> Glad to hear it.Do you have any schedule for this? Just to mention, it's not only about implementing refreshPassword in JdbcConnectionProvider, but also redesigning JdbcConnectionConfig. DSQL doesn't require a username and password; instead, it creates a password using ak/sk or profile combined with region and cluster endpoint. Therefore, its parameter requirements for JdbcConnectionConfig are different, potentially changing the original parameter requirements, such as making username and password mandatory in the original design

Yes, this indeed different. How about implements jdbc sink and overwrite key method to let this works?

cloud456 commented:
Initially, I did consider using JDBC to implement the DSQL sink functionality. However, due to DSQL's unique characteristics that differ from other databases, implementing it based on JDBC might alter existing logic and affect other databases. This would require investing significant time in testing and verification to ensure correct logic without impacting other database functionalities. Currently, I don't have much time to invest in this approach, though perhaps it could be done in the future. And this DSQL sink connector I've submitted is implemented based on real user migration needs, and I hope it can be merged. Thank you!

Hisoka-X reviewed (commented):
Can you provide a screenshot of it running normally?

Hisoka-X commented on seatunnel-connectors-v2/connector-dsql/README.md:
```diff
@@ -0,0 +1,255 @@
+# AWS DSQL Sink Connector
```
We don't need README. We need docs in https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2/sink

Hisoka-X commented:
Also please check these TODO 
<img width="711" height="171" alt="image" src="https://github.com/user-attachments/assets/29722a81-7884-4002-8271-9b1a46e158b1" />

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

cloud456 committed:
Merge branch 'apache:dev' into dev

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

cloud456 committed:
Merge branch 'apache:dev' into dev

cloud456 committed:
Merge branch 'apache:dev' into dev

cloud456 committed:
Merge branch 'apache:dev' into dev

cloud456 commented:
Thank you for your reminder. I have made the modifications as requested. Please help review it again. Additionally, since the DSQL environment cannot be simulated through Docker,so I haven't written corresponding test cases. I've provided two screenshots of the DSQL sink connector during operation - one from the initial state and another showing a delete operation. I'm wondering if these can be used to verify its effectiveness, or if you could provide other suggestions on how to verify that it works properly.
<img width="2546" height="948" alt="snapshot for initial" src="https://github.com/user-attachments/assets/b1fccd2a-4197-4905-8f51-47ccc8adad09" />
![delete recode](https://github.com/user-attachments/assets/936b2bb8-fafc-4a5c-a3e8-529c87298aec)

Hisoka-X commented:
Could you provide the screenshots that write data into DSQL success? For example, the job end with finished status, and DSQL had data to query.

Hisoka-X commented:
> How about implements jdbc sink and overwrite key method to let this works

We should consider this option carefully. The core difference is only the authentication method. We are fully capable of abstracting out a public interface for DSQL to implement.

cloud456 commented:
Hi，Here are two screenshots. The first picture is a screenshot of the log when the task is successfully completed. The second picture shows the comparison between the source table and the target table, hoping to prove its effectiveness.
<img width="2426" height="1168" alt="finished" src="https://github.com/user-attachments/assets/abf8b9bc-bcf2-4016-a12a-866639746009" />

<img width="1830" height="989" alt="Snipaste_2025-09-12_14-15-49" src="https://github.com/user-attachments/assets/4c9fc4f7-c170-4bbb-8837-64578083e33b" />

cloud456 committed:
Merge branch 'apache:dev' into dev

cloud456 committed:
Merge branch 'apache:dev' into dev

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726  delete dsql connector code

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726  implement base on jdbc

cloud456 committed:
Merge branch 'apache:dev' into dev

cloud456 committed:
Merge branch 'apache:dev' into dev

cloud456 committed:
Merge branch 'apache:dev' into dev

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

cloud456 committed:
Merge branch 'apache:dev' into dev

cloud456 committed:
Merge branch 'apache:dev' into dev

cloud456 committed:
Merge branch 'apache:dev' into dev

cloud456 committed:
Merge branch 'apache:dev' into dev

cloud456 commented:
I have used jdbc to implement the sink function of dsql.The following two screenshots are the results of running.

<img width="1780" height="784" alt="mysql_to_dsql_jdbc_finish" src="https://github.com/user-attachments/assets/b253819b-3b2e-4ac0-96cd-281865bb0f67" />
<img width="1322" height="701" alt="source target compare" src="https://github.com/user-attachments/assets/083b7354-fc14-4abf-86d8-38a87b2808ce" />

davidzollo commented on seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/dsql/DsqlJdbcConnectionProvider.java:
```diff
@@ -0,0 +1,105 @@
+                        .build();
+
+        if ("admin".equals(jdbcConfig.getUsername())) {
+            return dsqlUtilities.generateDbConnectAdminAuthToken(tokenGenerator);
```
It can't work fine because jdbcConfig.getUsername() return type `Optional<String>`

davidzollo commented:
Considering tokens that rotate every 15 minutes. `DsqlJdbcConnectionProvider` cannot synchronously refresh all connections, leading to batch failures after expiration; complex refresh and eviction logic is required.

In summary, I suggest not using `ConnectionPool`. By the way, please add a chinese doc.

cloud456 committed:
Merge branch 'apache:dev' into dev

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

cloud456 commented:
> Considering tokens that rotate every 15 minutes. `DsqlJdbcConnectionProvider` cannot synchronously refresh all connections, leading to batch failures after expiration; complex refresh and eviction logic is required.
> 
> In summary, I suggest not using `ConnectionPool`. By the way, please add a chinese doc.

Hi , I added the Chinese doc. In addition, regarding token refresh, I added a Schedule to `DsqlConnectionPoolManager`, and the token was refreshed every ten minutes, and the test results showed normal.

davidzollo commented:
> > Considering tokens that rotate every 15 minutes. `DsqlJdbcConnectionProvider` cannot synchronously refresh all connections, leading to batch failures after expiration; complex refresh and eviction logic is required.
> > In summary, I suggest not using `ConnectionPool`. By the way, please add a chinese doc.
> 
> Hi , I added the Chinese doc. In addition, regarding token refresh, I added a Schedule to `DsqlConnectionPoolManager`, and the token was refreshed every ten minutes, and the test results showed normal.

tokenRefreshExecutor should shutdown() when close

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

cloud456 committed:
Merge branch 'apache:dev' into dev

cloud456 committed:
Merge branch 'apache:dev' into dev

cloud456 committed:
Merge branch 'apache:dev' into dev

Commit:
[Feature][Connector-V2] Support sink connector for AWS DSQL #9726

cloud456 committed:
Merge branch 'apache:dev' into dev

davidzollo commented:
+1 if CI passed

cloud456 committed:
Merge branch 'apache:dev' into dev

davidzollo committed:
Merge branch 'dev' into dev

davidzollo reviewed (approved):
+1

chl-wxp commented:
Is there any way to add some test cases？

cloud456 commented:
> Is there any way to add some test cases？

Sorry，because Aurora DSQL is a cloud service, it can't simulate the environment through docker, so I can't provide more test cases at present. I have provided some screenshots after running above, hoping to help.

corgy-w commented:
@cloud456 Hi ,You might be able to add some @Disabled test cases
